File: src/oss_pipeline/contest.py
"""Phase B: is this issue already being worked on?

The bias here is deliberately conservative: a PR is treated as ACTIVE unless it
is *provably* abandoned. Opening a competing PR against someone's live work is
the single most reliably resented thing an outside contributor can do, and the
cost lands on the user's account, so ambiguity resolves to "leave it alone".

Nothing here judges the quality of anyone's code. Classification uses dates,
labels and review states only. Quality ranking exists solely to prioritise
*within* the stale bucket -- never to justify contesting live work.
"""

from __future__ import annotations

from datetime import datetime, timezone

from . import ghapi, policy
from .models import Candidate, Contest, PRSignal

PR_DETAIL = """
query($owner:String!, $name:String!, $number:Int!) {
  repository(owner:$owner, name:$name) {
    pullRequest(number:$number) {
      number url isDraft state createdAt updatedAt
      author { login }
      labels(first:20) { nodes { name } }
      commits(last:1) {
        nodes { commit { committedDate statusCheckRollup { state } } }
      }
      comments(last:30) { nodes { createdAt author { login } } }
      reviews(last:30) { nodes { state createdAt author { login } } }
    }
  }
}
"""

STALE_LABEL_HINTS = ("stale", "abandoned", "inactive", "no-response", "needs-rebase")


def _days_since(iso: str | None) -> int | None:
    if not iso:
        return None
    when = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    return (datetime.now(timezone.utc) - when).days


def _signal(repo: str, number: int) -> PRSignal:
    owner, name = repo.split("/")
    pr = ghapi.graphql(PR_DETAIL, owner=owner, name=name, number=number)[
        "repository"]["pullRequest"]

    author = (pr.get("author") or {}).get("login", "")
    commit_nodes = pr["commits"]["nodes"]
    commit = commit_nodes[0]["commit"] if commit_nodes else {}
    rollup = (commit.get("statusCheckRollup") or {}).get("state")

    author_comments = [
        c["createdAt"] for c in pr["comments"]["nodes"]
        if (c.get("author") or {}).get("login") == author
    ]
    changes_requested = [
        r["createdAt"] for r in pr["reviews"]["nodes"]
        if r.get("state") == "CHANGES_REQUESTED"
    ]
    labels = [l["name"].lower() for l in pr["labels"]["nodes"]]

    return PRSignal(
        number=pr["number"],
        url=pr["url"],
        author=author,
        is_draft=pr["isDraft"],
        days_since_commit=_days_since(commit.get("committedDate")),
        days_since_author_comment=_days_since(max(author_comments) if author_comments else None),
        days_since_changes_requested=_days_since(max(changes_requested) if changes_requested else None),
        has_stale_label=any(h in l for l in labels for h in STALE_LABEL_HINTS),
        checks_failing=rollup in ("FAILURE", "ERROR"),
    )


def _classify_one(sig: PRSignal) -> tuple[Contest, list[str]]:
    st = policy.policy()["staleness"]
    reasons: list[str] = []

    # Any recent author activity => live work. Checked first and wins outright.
    recent = [
        d for d in (sig.days_since_commit, sig.days_since_author_comment)
        if d is not None and d <= st["active_pr_days"]
    ]
    if recent:
        return Contest.ACTIVE_PR, [
            f"author active {min(recent)}d ago (within {st['active_pr_days']}d window)"
        ]

    if sig.has_stale_label:
        reasons.append("carries a stale/abandoned label")
    if (sig.days_since_commit or 0) > st["author_silent_days"]:
        reasons.append(f"no commit for {sig.days_since_commit}d")
    if (sig.days_since_changes_requested or 0) > st["changes_requested_days"]:
        unanswered = (
            sig.days_since_author_comment is None
            or sig.days_since_author_comment > sig.days_since_changes_requested
        )
        if unanswered:
            reasons.append(
                f"changes requested {sig.days_since_changes_requested}d ago, unanswered"
            )
    if sig.checks_failing and (sig.days_since_commit or 0) > st["ci_red_untouched_days"]:
        reasons.append(f"CI red and untouched for {sig.days_since_commit}d")

    if reasons:
        return Contest.STALE_PR, reasons

    # Between the windows with no positive staleness signal: do not contest.
    return Contest.ACTIVE_PR, [
        "no activity in the active window, but no staleness signal either "
        "-- treated as live (conservative default)"
    ]


def classify(cand: Candidate) -> tuple[Contest, PRSignal | None]:
    prs = getattr(cand, "_linked_prs", None)
    if prs is None:
        prs = []
    if not prs:
        return Contest.NO_PR, None

    signals = []
    for pr in prs:
        try:
            signals.append(_signal(cand.repo, pr["number"]))
        except ghapi.GhError as exc:
            # Unknown state is not evidence of abandonment.
            print(f"    {cand.slug}: PR #{pr['number']} unreadable ({str(exc)[:50]})"
                  f" -- treating issue as contested")
            return Contest.ACTIVE_PR, None

    verdicts = [_classify_one(s) for s in signals]
    # One live PR is enough to make the issue off limits.
    for (verdict, reasons), sig in zip(verdicts, signals):
        if verdict is Contest.ACTIVE_PR:
            sig.reasons = reasons
            return Contest.ACTIVE_PR, sig

    # All stale: surface the one that is most abandoned (longest silence).
    best = max(
        zip(signals, verdicts),
        key=lambda pair: pair[0].days_since_commit or 0,
    )
    best[0].reasons = best[1][1]
    return Contest.STALE_PR, best[0]

File: internal/contest/contest.go
// Package contest answers one question: is this issue already being worked on?
//
// The bias is deliberately conservative. A pull request is treated as ACTIVE
// unless it is provably abandoned, because opening a competing pull request
// against someone's live work is the most reliably resented thing an outside
// contributor can do, and the cost lands on the user's account rather than on
// this code. Ambiguity resolves to "leave it alone".
//
// Nothing here judges anyone's code. Classification uses dates, labels and
// review states only. The ranking of stale pull requests exists to choose
// between abandoned ones, never to justify contesting live work.
package contest

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/vimalyad/osspipeline/internal/model"
	"github.com/vimalyad/osspipeline/internal/policy"
)

// API is the GraphQL surface this package needs.
type API interface {
	GraphQL(ctx context.Context, query string, vars map[string]any, v any) error
}

// staleLabelHints are substrings maintainers use to mark a pull request as
// no longer moving.
var staleLabelHints = []string{"stale", "abandoned", "inactive", "no-response", "needs-rebase"}

const prDetailQuery = `
query($owner:String!, $name:String!, $number:Int!) {
  repository(owner:$owner, name:$name) {
    pullRequest(number:$number) {
      number url isDraft state createdAt updatedAt
      author { login }
      labels(first:20) { nodes { name } }
      commits(last:1) {
        nodes { commit { committedDate statusCheckRollup { state } } }
      }
      comments(last:30) { nodes { createdAt author { login } } }
      reviews(last:30) { nodes { state createdAt author { login } } }
    }
  }
}`

type prDetail struct {
	Repository struct {
		PullRequest struct {
			Number    int    `json:"number"`
			URL       string `json:"url"`
			IsDraft   bool   `json:"isDraft"`
			State     string `json:"state"`
			CreatedAt string `json:"createdAt"`
			UpdatedAt string `json:"updatedAt"`
			Author    *struct {
				Login string `json:"login"`
			} `json:"author"`
			Labels struct {
				Nodes []struct {
					Name string `json:"name"`
				} `json:"nodes"`
			} `json:"labels"`
			Commits struct {
				Nodes []struct {
					Commit struct {
						CommittedDate     string `json:"committedDate"`
						StatusCheckRollup *struct {
							State string `json:"state"`
						} `json:"statusCheckRollup"`
					} `json:"commit"`
				} `json:"nodes"`
			} `json:"commits"`
			Comments struct {
				Nodes []struct {
					CreatedAt string `json:"createdAt"`
					Author    *struct {
						Login string `json:"login"`
					} `json:"author"`
				} `json:"nodes"`
			} `json:"comments"`
			Reviews struct {
				Nodes []struct {
					State     string `json:"state"`
					CreatedAt string `json:"createdAt"`
				} `json:"nodes"`
			} `json:"reviews"`
		} `json:"pullRequest"`
	} `json:"repository"`
}

// LinkedPR is a pull request that references the issue.
type LinkedPR struct {
	Number int
	URL    string
}

// Signal fetches the auditable facts about one pull request.
func Signal(ctx context.Context, api API, repo string, number int, now time.Time) (*model.PRSignal, error) {
	owner, name, ok := strings.Cut(repo, "/")
	if !ok {
		return nil, fmt.Errorf("contest: %q is not owner/name", repo)
	}
	var resp prDetail
	if err := api.GraphQL(ctx, prDetailQuery, map[string]any{
		"owner": owner, "name": name, "number": number,
	}, &resp); err != nil {
		return nil, fmt.Errorf("contest %s#%d: %w", repo, number, err)
	}
	pr := resp.Repository.PullRequest

	author := ""
	if pr.Author != nil {
		author = pr.Author.Login
	}

	var committed string
	var rollup string
	if len(pr.Commits.Nodes) > 0 {
		c := pr.Commits.Nodes[0].Commit
		committed = c.CommittedDate
		if c.StatusCheckRollup != nil {
			rollup = c.StatusCheckRollup.State
		}
	}

	var lastAuthorComment string
	for _, c := range pr.Comments.Nodes {
		if c.Author == nil || c.Author.Login != author || author == "" {
			continue
		}
		if c.CreatedAt > lastAuthorComment {
			lastAuthorComment = c.CreatedAt
		}
	}
	var lastChangesRequested string
	for _, r := range pr.Reviews.Nodes {
		if r.State == "CHANGES_REQUESTED" && r.CreatedAt > lastChangesRequested {
			lastChangesRequested = r.CreatedAt
		}
	}

	var labels []string
	for _, l := range pr.Labels.Nodes {
		labels = append(labels, strings.ToLower(l.Name))
	}

	return &model.PRSignal{
		Number:                   pr.Number,
		URL:                      pr.URL,
		Author:                   author,
		IsDraft:                  pr.IsDraft,
		DaysSinceCommit:          daysSince(committed, now),
		DaysSinceAuthorComment:   daysSince(lastAuthorComment, now),
		DaysSinceChangesReqested: daysSince(lastChangesRequested, now),
		HasStaleLabel:            hasStaleLabel(labels),
		ChecksFailing:            rollup == "FAILURE" || rollup == "ERROR",
	}, nil
}

// ClassifyOne decides whether a single pull request makes an issue off limits.
func ClassifyOne(sig *model.PRSignal, st policy.Staleness) (model.Contest, []string) {
	// Any recent author activity means live work. Checked first and wins
	// outright: nothing below can talk us into contesting it.
	best := -1
	for _, d := range []*int{sig.DaysSinceCommit, sig.DaysSinceAuthorComment} {
		if d != nil && *d <= st.ActivePRDays && (best < 0 || *d < best) {
			best = *d
		}
	}
	if best >= 0 {
		return model.ContestActivePR, []string{fmt.Sprintf(
			"author active %dd ago (within %dd window)", best, st.ActivePRDays)}
	}

	var reasons []string
	if sig.HasStaleLabel {
		reasons = append(reasons, "carries a stale/abandoned label")
	}
	if sig.DaysSinceCommit != nil && *sig.DaysSinceCommit > st.AuthorSilentDays {
		reasons = append(reasons, fmt.Sprintf("no commit for %dd", *sig.DaysSinceCommit))
	}
	if sig.DaysSinceChangesReqested != nil && *sig.DaysSinceChangesReqested > st.ChangesRequestedDays {
		// Unanswered means the author has not spoken since the request. An
		// author who replied and is waiting on the maintainer is not stalling.
		unanswered := sig.DaysSinceAuthorComment == nil ||
			*sig.DaysSinceAuthorComment > *sig.DaysSinceChangesReqested
		if unanswered {
			reasons = append(reasons, fmt.Sprintf(
				"changes requested %dd ago, unanswered", *sig.DaysSinceChangesReqested))
		}
	}
	if sig.ChecksFailing && sig.DaysSinceCommit != nil && *sig.DaysSinceCommit > st.CIRedUntouchedDays {
		reasons = append(reasons, fmt.Sprintf(
			"CI red and untouched for %dd", *sig.DaysSinceCommit))
	}
	if len(reasons) > 0 {
		return model.ContestStalePR, reasons
	}

	// Between the windows with no positive staleness signal. Silence is not
	// evidence of abandonment, so this stays off limits.
	return model.ContestActivePR, []string{
		"no activity in the active window, but no staleness signal either -- " +
			"treated as live (conservative default)"}
}

// Classify decides for a whole issue, given every pull request linked to it.
func Classify(ctx context.Context, api API, repo string, prs []LinkedPR, st policy.Staleness, now time.Time) (model.Contest, *model.PRSignal, error) {
	if len(prs) == 0 {
		return model.ContestNoPR, nil, nil
	}

	type scored struct {
		sig     *model.PRSignal
		verdict model.Contest
		reasons []string
	}
	var all []scored
	for _, pr := range prs {
		sig, err := Signal(ctx, api, repo, pr.Number, now)
		if err != nil {
			// Unknown state is not evidence of abandonment. An unreadable
			// pull request makes the issue contested, not available.
			return model.ContestActivePR, nil, nil
		}
		v, reasons := ClassifyOne(sig, st)
		all = append(all, scored{sig, v, reasons})
	}

	// One live pull request is enough to put the issue off limits.
	for _, s := range all {
		if s.verdict == model.ContestActivePR {
			s.sig.Reasons = s.reasons
			return model.ContestActivePR, s.sig, nil
		}
	}

	// All stale: surface the most abandoned one, by longest silence.
	bestIdx := 0
	for i, s := range all {
		if days(s.sig.DaysSinceCommit) > days(all[bestIdx].sig.DaysSinceCommit) {
			bestIdx = i
		}
	}
	all[bestIdx].sig.Reasons = all[bestIdx].reasons
	return model.ContestStalePR, all[bestIdx].sig, nil
}

func days(p *int) int {
	if p == nil {
		return 0
	}
	return *p
}

// daysSince returns nil for an absent or unparseable timestamp, which every
// caller must treat as "unknown" rather than as zero. A missing date read as
// "0 days ago" would make an abandoned pull request look active; read as
// "very old" it would make a live one look abandoned. Only nil is honest.
func daysSince(iso string, now time.Time) *int {
	if strings.TrimSpace(iso) == "" {
		return nil
	}
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		return nil
	}
	d := int(now.UTC().Sub(t).Hours() / 24)
	if d < 0 {
		d = 0
	}
	return &d
}

func hasStaleLabel(labels []string) bool {
	for _, l := range labels {
		for _, h := range staleLabelHints {
			if strings.Contains(l, h) {
				return true
			}
		}
	}
	return false
}
